cmd/data-usage.go

```diff
@@ -0,0 +1,198 @@
+	go runDataUsageInfoUpdateRoutine()
+}
+
+func runDataUsageInfoUpdateRoutine() {
+	// Wait until the object layer is ready
+	var objAPI ObjectLayer
+	for {
+		objAPI = newObjectLayerWithoutSafeModeFn()
+		if objAPI == nil {
+			time.Sleep(time.Second)
+			continue
+		}
+		break
+	}
+
+	ctx := context.Background()
+
+	switch v := objAPI.(type) {
+	case *xlZones:
+		runDataUsageInfoForXLZones(ctx, v)
+	case *FSObjects:
+		runDataUsageInfoForFS(ctx, v)
+	default:
+		return
+	}
+}
+
+func runDataUsageInfoForFS(ctx context.Context, fsObj *FSObjects) {
+	for {
+		// Get data usage info of the FS Object
+		usageInfo := fsObj.crawlAndGetDataUsageInfo(ctx)
+		// Save the data usage in the disk
+		err := storeDataUsageInBackend(ctx, fsObj, usageInfo)
+		if err != nil {
+			logger.LogIf(ctx, err)
+		}
+		// Wait until the next crawl interval
+		time.Sleep(dataUsageCrawlInterval)
+	}
+}
+
+func runDataUsageInfoForXLZones(ctx context.Context, zones *xlZones) {
+	for {
+		locker := zones.NewNSLock(ctx, minioMetaBucket, "leader-data-usage-info")
+		err := locker.GetLock(newDynamicTimeout(time.Millisecond, time.Millisecond))
+		if err != nil {
+			time.Sleep(5 * time.Minute)
+			continue
+		}
+		// Break without locking
+		break
+	}
+
+	for {
+		usageInfo := zonesCrawlAndGetDataUsage(ctx, zones)
+		err := storeDataUsageInBackend(ctx, zones, usageInfo)
+		if err != nil {
+			logger.LogIf(ctx, err)
+		}
+
+		time.Sleep(dataUsageCrawlInterval)
+	}
+
+}
+
+func zonesCrawlAndGetDataUsage(ctx context.Context, zones *xlZones) DataUsageInfo {
+	// Calculate the aggregated data usage, meaning
+	// accumulate data usage of all zones.
+	var aggregatedDataUsageInfo = DataUsageInfo{
+		ObjectsSizesHistogram: make(map[string]uint64),
+		BucketsSizes:          make(map[string]uint64),
+	}
+
+	addUpUsageInfo := func(dataUsageInfo DataUsageInfo) {
+		aggregatedDataUsageInfo.ObjectsCount += dataUsageInfo.ObjectsCount
+		aggregatedDataUsageInfo.ObjectsTotalSize += dataUsageInfo.ObjectsTotalSize
+		if aggregatedDataUsageInfo.BucketsCount < dataUsageInfo.BucketsCount {
+			aggregatedDataUsageInfo.BucketsCount = dataUsageInfo.BucketsCount
+		}
+		for k, v := range dataUsageInfo.ObjectsSizesHistogram {
+			aggregatedDataUsageInfo.ObjectsSizesHistogram[k] += v
+		}
+		for k, v := range dataUsageInfo.BucketsSizes {
+			aggregatedDataUsageInfo.BucketsSizes[k] += v
+		}
+	}
+
+	for _, z := range zones.zones {
+		for _, xl := range z.sets {
+			var randomDisks []StorageAPI
+			for _, d := range xl.getLoadBalancedDisks() {
+				if d == nil || !d.IsOnline() {
+					continue
+				}
+				if len(randomDisks) > 3 {
+					break
+				}
+				randomDisks = append(randomDisks, d)
+			}
+
+			var dataUsageResult = make([]DataUsageInfo, len(randomDisks))
+
+			var wg sync.WaitGroup
+			for i := 0; i < len(randomDisks); i++ {
+				wg.Add(1)
+				go func(index int, disk StorageAPI) {
+					defer wg.Done()
+					var err error
+					dataUsageResult[index], err = disk.CrawlAndGetDataUsage()
```

Comment:
note that crawl should happen across sets of all zones in parallel

Reply:
> note that crawl should happen across sets of all zones in parallel

Done